### DISCUSSION TITLE
[Question]: How can I get learnt clause while solving formula in cvc5?

### DISCUSSION THREAD

**Discussion by hagozaebii** [cvc5/cvc5#11221] [Q&A] [answered]
Hello,

I'm interested in observing the clauses learned through the CDCL algorithm in real-time during formula solving. Is there a way to access or visualize these learned clauses as the SMT solver progresses in cvc5? Specifically, I'm curious if there's a verbose mode or debug option that could output this information, or if there's an API of python that allows monitoring of the clause learning process. Any insights on how to track the evolution of the clause database during solving would be greatly appreciated. 

Thank you for your expertise on this matter.

**Comment by ajreynol** [comment, marked as answer]
Hello, 

Yes, we have a "Plugin" feature in our API.

See this call: https://github.com/cvc5/cvc5/blob/main/include/cvc5/cvc5.h#L3633

If you register a Plugin to a Solver, it will be notified of when sat clauses and theory lemmas are learned during solving.

You can also use `-o lemmas` to output theory lemmas as they are generated during solving.  For more information on this trace and others, you can see this blog post I wrote recently: https://cvc5.github.io/blog/2024/04/15/interfaces-for-understanding-cvc5.html

**Comment by hagozaebii** [comment]
@ajreynol  Thanks for your kindness !
